jquery.flot.touchNavigate.js

```diff
@@ -72,31 +74,90 @@ can set the default in the options. */
             }
         }
 
+        function isPinchEvent(e) {
+            return e.touches && e.touches.length === 2;
+        }
+
         function onDragStart(e) {
             e.stopPropagation();
             e.preventDefault();
-            plot.getPlaceholder().css('cursor', plot.getOptions().pan.cursor);
-            startPageX = e.touches[0].pageX || e.touches[0].clientX;
-            startPageY = e.touches[0].pageY || e.touches[0].clientY;
-            plotState = plot.navigationState();
+            scaling = isPinchEvent(e);
+
+            if (!isPinchEvent(e)) {
+                scaling = false;
 
-            saveNavigationData(plot, e);
+                plot.getPlaceholder().css('cursor', plot.getOptions().pan.cursor);
+                startPageX = e.touches[0].pageX || e.touches[0].clientX;
+                startPageY = e.touches[0].pageY || e.touches[0].clientY;
+                plotState = plot.navigationState();
+
+                saveNavigationData(plot, e);
+            } else {
+                scaling = true;
+                prevDist = pinchDistance(e);
+            }
         }
 
         function onDrag(e) {
             e.stopPropagation();
             e.preventDefault();
 
-            plot.smartPan({
-                x: startPageX - (e.touches[0].pageX || e.touches[0].clientX),
-                y: startPageY - (e.touches[0].pageY || e.touches[0].clientY)
-            }, plotState);
-
-            saveNavigationData(plot, e);
+            scaling = isPinchEvent(e);
+
+            if (!scaling) {
+                return false;
+                plot.smartPan({
+                    x: startPageX - (e.touches[0].pageX || e.touches[0].clientX),
+                    y: startPageY - (e.touches[0].pageY || e.touches[0].clientY)
+                }, plotState);
+
+                saveNavigationData(plot, e);
+            } else if (isPinchEvent(e)) {
+                var dist = pinchDistance(e);
+                if (Math.abs(100 * (dist - prevDist) / prevDist) > 5) {
+                    onZoomPinch(e, dist < prevDist);
+                    prevDist = dist;
+                }
+            }
         }
 
-        function onDragEnd() {
+        function onDragEnd(e) {
             // not sure if this should do anything
+            if (scaling) {
+                if (isPinchEvent(e)) {
+                    //var dist = pinchDistance(e);
+                    scaling = false;
+                    prevDist = null
+                }
+            }
+        }
+
+        function onZoomPinch(e, zoomOut) {
+            var c = plot.offset(),
+                pageX = (e.touches[0].clientX + e.touches[1].clientX) / 2,
+                pageY = (e.touches[0].clientY + e.touches[1].clientY) / 2;
+
+            c.left = pageX - c.left; //pageX
+            c.top = pageY - c.top; //pageY
```

Comment:
Let's use a variable for offset and a different one for center. Longer than 1 char names are preferred.